usenix89/Graphics/GKSdemo/mongo.man
.TH MONGO 1 local
.SH NAME
.PP
mongo - a simple interactive graphing program
.SH DESCRIPTION
.PP
Mongo allows the user to produce simple graphs composed of
lines, points and text, together with labeled axes. Input is
taken from files which have several columns (numeric fields
separated by white space) per line; one may plot the
contents of any column versus any other.
.PP
Hardcopy may be produced in two ways: either by screen dump 
or by direct output to a printer.
The first method yields graphs with an aspect ratio 
slightly more cramped vertically (at least on the AT&T 7300)
than the screen; the second gives a much more even aspect ratio.
.SH Commands
There are many commands; a brief description of each one follows:
.IP quit 20
Quit the program, clear the screen and set the window back to its
initial size. Any hardcopy plots produced via 
.I update
will be made at this time.
.IP clear 20
Clear the graphing area of the screen and any other attached graphics
device such as the printer. A formfeed character will be sent to the
printer and the page will be queued for printing upon exiting the program.
.IP "file data" 20
Open the file 
.I data
for data input. If it does not exist, the program will complain.
.IP "lines n1 n2" 20
Limit the lines of the file from which data will be read to start
at line 
.I n1
and end at line 
.I n2. 
Line numbers start at 1, not 0.
.IP "xc n" 20
Take data for the x coordinates from column number 
.I n 
in the file.
Column separators are any white space. Non-numeric data causes
bad things to happen, and will probably screw up all following
operations.
.IP "yc n" 20
Same as 
.I xc
above, but for the y coordinates.
.IP "limits [ x1 x2 y1 y2 ]" 20
If four arguments are present, set the limits of the plot to the
given values: the x-values range from a lower limit of 
.I x1 
to 
.I x2,
the y-values from 
.I y1 
to 
.I y2. 
If no arguments are given, set the
limits to the extremes of the data (so that all points should
appear in the plot).
.IP points 20
Place the current point markers (set by
.I ptype) 
at each one of the data points which are within the 
screen boundaries.
.IP "ptype [ value ]" 20
If an argument is given, set the current point type to the given
.I value; 
with no argument, report the current point type.
Possible types are POINT, PLUS, ASTERISK, CIRCLE, CROSS, DIAMOND.
The default is PLUS.
.IP connect 20
Draw lines connecting the current data points, clipping lines to
off-screen points at the plot boundaries. The type of line may
be selected via 
.I ltype.
.IP "ltype [ value ]" 20
If an argument is given, set the current line type to the given
.I value;
with no argument, report the current line type. Possible line types
are SOLID, LONG_DASH, DOT, DOT_DASH, MEDIUM_DASH and DASH_2_DOTS.
DOT is the default.
.IP box 20
Draw four solid lines on the screen, outlining the plot area.
.IP "axes [ xscale yscale ]" 20
Similar to 
.I box
above, but the axes (x on the bottom, y on the left) are tick-marked 
and labeled with the values of the respective variable. Each value
is in the form of a single digit, a decimal point 
and two decimal places; the scale factor is placed on the far end of 
the axis. If arguments
.I xscale
and
.I yscale
are given, they are used to scale the x axis and the y axis, respectively.
Otherwise, the axes are scaled by the program with reasonable values.
.IP "xaxis label" 20
Place the text string 
.I label,
which may be several words, horizontally below the x axis.
.IP "yaxis label" 20
Place the text string
.I label
vertically (reading upwards) to the left of the y axis.
.IP "title label" 20
Place the string
.I label
above the graph and centered on it.
.IP "position [ xp yp ]" 20
If two numeric arguments are given, move 
the current text position to the location 
.I "(xp, yp)"
on the graph. Both 
.I xp
and
.I yp
must be in the current graph coordinates (i.e. if x runs from 1 to 6
and y from 4 to 12, one might say
.I "position 2 11"
to move into the upper left-hand corner of the plot.
If no arguments, report the current text position.
.IP "label text" 20
Place the string
.I text,
which may be more than one word, at the current text position on the
graph. If the current position is not inside the plot boundaries,
do nothing.
.IP reverse 20
Applies only to the graphics window of the screen. Reverse the color of
the screen so that what was previously black is white and the white is
black.
.PP
The following commands are used only with attached external hardcopy devices
such as printer.
.IP printer 20
Open the printer workstation for hardcopy plots. All subsequent commands
issued will be sent to the printer as well as the screen until
.I noprinter
command is issued.
.IP noprinter 20
Close the printer workstation.
.IP dump 20
Take a screen dump. The content of the graphics window is sent to the
printer.
.IP update 20
All graphics commands which have been given since the last
.I update
or
.I clear
are sent to the external device; thus, successive calls to 
.I update 
will separate hardcopy plots on the external device.
Note, however, that actual hardcopy is not produced at this
time, but only queued; actual printing takes place after the
workstation is closed via the 
.I quit
command.

.SH BUGS
.PP
The
.I dump
command is device dependent.
.PP
Hardcopy is not produced immediately after each
.I update
or
.I clear,
but only after the workstation is closed via
.I quit.
This seems to be a GSS_Drivers "feature," but it might be possible
to flush output to external devices via some system call - it certainly
would be nice.
.PP
In the current version, the windows, once created, cannot be moved or
resized.
.SH AUTHORS
.PP
Ann Takata and Michael Richmond. Send mail to astrovax!richmon.



usenix89/Graphics/GKSdemo/mongo.c
#include <fcntl.h>
#include <tam.h>
#include <ctype.h>
#include "mongo.h"


INT16 display[] = {
	1,	/* equal axis ratios */
	1,	/* solid line type */
	1,	/* line color index to use - white */
	3,	/* marker type - star */
	1,	/* marker color index - white */
	1,	/* text font to use - normal */
	1,	/* text color - white */
	0,	/* fill interior style - hollow */
	0,	/* fill style index - plain */
	1,	/* fill color index - white */
	1,	/* send device messages to screen */
	'D', 'I', 'S', 'P', 'L', 'A', 'Y', ' '	/* device name */
};

char *line_type[] = {
	"SOLID",
	"LONG_DASH",
	"DOT",
	"DASH_DOT",
	"MEDIUM_DASH",
	"DASH_2_DOTS"
};

char *point_type[] = {
	"POINT",
	"PLUS",
	"ASTERISK",
	"CIRCLE",
	"CROSS",
	"DIAMOND"
};


main(argc, argv)
int argc; 
char *argv[];
{
	int i, j, k, num1, num2;
	char *gets(), cmd_buf[50], command[10];
	struct nlist *lookup(), *np;

	init();

	while (1) {
		command[0] = '\0';
		wselect (text_w);
		wprintf(text_w, "> ");
		wgets(text_w, cmd_buf);
		sscanf(cmd_buf, "%s", command);
		if ((np = lookup(command)) == NULL) {
			wprintf(text_w, "%s ?\n", command);
			continue;
		}
		(*(np->func))(cmd_buf);
	}
}


	/* set up two windows, one for the graphics, one for text and
	   stuff. select the text one, so that it is on top */

init()
{
	short placex = 1;
	short placey = 0;
	short height, width;
	unsigned short flags;
	struct utdata utd;
	struct nlist *install();
	INT16 dev_info[66], text_attrib[10], err_report;
	int my_clear(), get_file(), get_xc(), get_yc(), get_limits();
	int points(), connect(), box(), quit(), ltype(), ptype();
	int lines(), position(), label(), reverse();
	int draw_axes(), xlabel(), ylabel(), title();
#ifdef PRINTER
	int my_update();
	int init_print(), end_print(), my_print();
#endif

	winit();

	/* create the graphics window */

	height = 24;
	width = 80;
	flags = NBORDER;
	graph_w = wcreate(placex, placey, height, width, flags);
	utd.ut_num = WTXTLABEL;
	strcpy(utd.ut_text, "Mongo Graph");
	ioctl(graph_w, WIOCSETTEXT, &utd);

	/* create the text window */

	height = 4;
	width = 30;
	flags = BORDRESIZE;
	text_w = wcreate(placex, placey, height, width, flags);
	utd.ut_num = WTXTLABEL;
	strcpy(utd.ut_text, "Mongo Commands");
	ioctl(text_w, WIOCSETTEXT, &utd);

	/* open graphics workstation */
	wselect(graph_w);
	err_report = v_opnwk(display, &dev_crt, dev_info);
	if (err_report != 0) {
		wselect(text_w);
		wprintf(text_w,"error opening screen: %d\n", vq_error());
		wexit();
	}

	/* set graph borders and text borders */
	vqt_attributes(dev_crt, text_attrib);
	T_HEIGHT = text_attrib[9];
	T_WIDTH = text_attrib[8];
	MINX = 8 * T_WIDTH;
	MAXX = dev_info[51] - T_WIDTH;
	MINY = 4 * T_HEIGHT;
	MAXY = dev_info[52] - 2 * T_HEIGHT;

	/* set some characteristics of the graphics window */
	cur_p_type = PLUS;
	cur_l_type = DOT;
	vsm_type(dev_crt, cur_p_type);		/* plus signs as points */
	vsl_type(dev_crt, cur_l_type);		/* dotted line */

	/* set a few other things */
	low_line = 1;
	high_line = AR_SIZE;

	/* initialize printer and plotter flags */
	PRINT = 0;
	PLOT = 0;

	/* install commands in the command table */
	table_init();
	install("clear", my_clear);
	install("file", get_file);
	install("xc", get_xc);
	install("yc", get_yc);
	install("lines", lines);
	install("limits", get_limits);
	install("ltype", ltype);
	install("ptype", ptype);
	install("points", points);
	install("connect", connect);
	install("box", box);
	install("position", position);
	install("label", label);
	install("reverse", reverse);
	install("quit", quit);
	install("axes", draw_axes);
	install("xaxis", xlabel);
	install("yaxis", ylabel);
	install("title", title);

#ifdef PRINTER
	install("update", my_update);
	install("dump", my_print);
	install("printer", init_print);
	install("noprinter", end_print);
#endif

}


	/* in the second two arguments, put the NDC coordinates corresponding 
	   to the world coordinates given in the first two arguments */

world_to_ndc(wx, wy, nx, ny)
double wx, wy, *nx, *ny;
{
	double fx, fy, cx, cy;

	fx = (MAXX - MINX) / (maxx - minx);
	fy = (MAXY - MINY) / (maxy - miny);
	cx = MINX - fx * minx;
	cy = MINY - fy * miny;

	*nx = wx * fx + cx;
	*ny = wy * fy + cy;
}


	/* set the type of line used in connect(); with no arguments,
	   prints the current type */

ltype(cmd_buf)
char *cmd_buf;
{
	int i;
	char dummy[7], type[20];

	if (sscanf(cmd_buf, "%s %s", dummy, type) < 2)
		wprintf(text_w, "%s\n", line_type[cur_l_type - 1]);
	else {
		for (i = 0; i <= (MAX_L_TYPE - MIN_L_TYPE); i++)
			if (strcmp(type, line_type[i]) == 0) {
				cur_l_type = i + MIN_L_TYPE;
				vsl_type(dev_crt, cur_l_type);

#ifdef PRINTER
				if (PRINT)
					vsl_type(dev_prt, cur_l_type);
#endif

			}
	}
}


	/* set the type of point used in points(); with no arguments,
	   prints the current type */

ptype(cmd_buf)
char *cmd_buf;
{
	int i;
	char dummy[7], type[20];

	if (sscanf(cmd_buf, "%s %s", dummy, type) < 2)
		wprintf(text_w, "%s\n", point_type[cur_p_type - 1]);
	else {
		for (i = 0; i <= (MAX_P_TYPE - MIN_P_TYPE); i++)
			if (strcmp(type, point_type[i]) == 0) {
				cur_p_type = i + MIN_P_TYPE;
				vsm_type(dev_crt, cur_p_type);

#ifdef PRINTER
				if (PRINT)
					vsm_type(dev_prt, cur_p_type);
#endif

			}
	}
}


	/* clear the graphics window and the printer, if appropriate */

my_clear()
{
	wselect (graph_w);
	v_clrwk(dev_crt);

#ifdef PRINTER
	if (PRINT)
		v_clrwk(dev_prt);
#endif

}


	/* make a hard copy */

#ifdef PRINTER
my_print ()
{
	int err;

	wselect(graph_w);
	err = v_hardcopy(dev_crt);
	if (err != 0) {
		wselect(text_w);
		wprintf(text_w, "error hardcopy\n");
	}
}
#endif


	/* update printer workstation */

#ifdef PRINTER
my_update()
{
	if (PRINT)
		v_updwk(dev_prt);
}
#endif


	/* quit */

quit()
{
	wselect(graph_w);
	v_clrwk(dev_crt);
	v_clswk(dev_crt);

#ifdef PRINTER
	if (PRINT)
		v_clswk(dev_prt);
#endif

	wexit();
}
